test/run-tests

```diff
@@ -0,0 +1,366 @@
+#!/usr/bin/python3 -u
+
+import datetime
+import glob
+import json
+import logging
+import os
+import random
+import requests
+import shlex
+import shutil
+import socket
+import subprocess
+import sys
+import tempfile
+import time
+import traceback
+import urllib.parse
+import urllib.request
+
+hostname = socket.gethostname()
+
+
+class Session(requests.Session):
+    """
+    A small extension for requests.Session that saves typing the host, calls
+    raise_for_status() on responses by default, and allows specifying a default
+    timeout.
+    """
+
+    def __init__(self, host, timeout=10):
+        super().__init__()
+        self.host = host
+        self.timeout = timeout
+
+    def post(self, path, json, check=True):
+        r = super().post(f'{self.host}/{path}', json=json, timeout=self.timeout)
+        if check:
+            r.raise_for_status()
+        return r
+
+    def get(self, path, check=True):
+        r = super().get(f'{self.host}/{path}', timeout=self.timeout)
+        if check:
+            r.raise_for_status()
+        return r
+
+
+class redirect_output:
+    """
+    A context manager that redirects stdout and stderr to a file.
+    """
+
+    def __init__(self, filename):
+        self.filename = filename
+
+    def __enter__(self):
+        self.f = open(self.filename, 'w', buffering=1)
+        sys.stdout.flush()
+        sys.stderr.flush()
+
+        self.oldout = sys.stdout
+        self.olderr = sys.stderr
+
+        sys.stdout = self.f
+        sys.stderr = self.f
+
+        return self.f
+
+    def __exit__(self, *exception):
+        self.f.close()
+
+        sys.stdout = self.oldout
+        sys.stderr = self.olderr
```

Comment:
The close should happen after setting stdout/stderr back. In addition there should be flush() calls here.

Reply:
Closing the file implies flushing, doesn't it? (Note that I am flushing `stdout` and `stderr` before redirecting, which was the main motivation of making this instead of using `contextlib.redirect_stdout`).

 Why is closing before setting them back preferable?